File: docs/comparison-tests/results-with-mcp/typescript/test-4-result.ts
/**
 * Test TS-4: Refactor
 * Refactor this function to be more readable and testable.
 *
 * Original code had issues:
 * - Too many responsibilities
 * - No type safety
 * - Hard to test (tightly coupled dependencies)
 * - Poor error handling
 * - Imperative style
 */

// ============================================================================
// Domain Types
// ============================================================================

interface OrderItem {
  readonly id: string;
  readonly name: string;
  readonly price: number;
  readonly qty: number;
}

interface Order {
  readonly id: string;
  readonly items: OrderItem[];
}

interface User {
  readonly id: string;
  readonly email: string;
  readonly balance: number;
}

interface PaymentResult {
  readonly success: boolean;
  readonly transactionId?: string;
  readonly error?: string;
}

// ============================================================================
// Result Types (Railway-Oriented Programming)
// ============================================================================

type OrderResult =
  | { readonly success: true; readonly orderId: string; readonly total: number }
  | { readonly success: false; readonly error: OrderError };

type OrderError =
  | { readonly type: 'INVALID_INPUT'; readonly message: string }
  | { readonly type: 'EMPTY_CART'; readonly message: string }
  | { readonly type: 'OUT_OF_STOCK'; readonly itemName: string; readonly message: string }
  | { readonly type: 'INSUFFICIENT_FUNDS'; readonly required: number; readonly available: number; readonly message: string }
  | { readonly type: 'PAYMENT_FAILED'; readonly message: string };

// ============================================================================
// Port Interfaces (Dependency Inversion)
// ============================================================================

interface Inventory {
  checkStock(itemId: string): number;
  decreaseStock(itemId: string, quantity: number): void;
}

interface PaymentGateway {
  charge(userId: string, amount: number): PaymentResult;
}

interface EmailService {
  sendOrderConfirmation(email: string, orderId: string): void;
}

// ============================================================================
// Order Validator
// ============================================================================

class OrderValidator {
  validateOrder(order: Order | null | undefined): OrderError | null {
    if (!order) {
      return { type: 'INVALID_INPUT', message: 'Order is required' };
    }
    if (!order.items || order.items.length === 0) {
      return { type: 'EMPTY_CART', message: 'Cart is empty' };
    }
    return null;
  }

  validateUser(user: User | null | undefined): OrderError | null {
    if (!user) {
      return { type: 'INVALID_INPUT', message: 'User is required' };
    }
    return null;
  }
}

// ============================================================================
// Order Calculator
// ============================================================================

class OrderCalculator {
  calculateTotal(items: OrderItem[]): number {
    return items.reduce((sum, item) => sum + item.price * item.qty, 0);
  }
}

// ============================================================================
// Stock Checker
// ============================================================================

class StockChecker {
  constructor(private readonly inventory: Inventory) {}

  checkAvailability(items: OrderItem[]): OrderError | null {
    for (const item of items) {
      const available = this.inventory.checkStock(item.id);
      if (available < item.qty) {
        return {
          type: 'OUT_OF_STOCK',
          itemName: item.name,
          message: `Item '${item.name}' is out of stock`,
        };
      }
    }
    return null;
  }

  reserveItems(items: OrderItem[]): void {
    for (const item of items) {
      this.inventory.decreaseStock(item.id, item.qty);
    }
  }
}

// ============================================================================
// Balance Checker
// ============================================================================

class BalanceChecker {
  checkBalance(user: User, total: number): OrderError | null {
    if (user.balance < total) {
      return {
        type: 'INSUFFICIENT_FUNDS',
        required: total,
        available: user.balance,
        message: 'Insufficient funds',
      };
    }
    return null;
  }
}

// ============================================================================
// Payment Processor
// ============================================================================

class PaymentProcessor {
  constructor(private readonly paymentGateway: PaymentGateway) {}

  processPayment(userId: string, total: number): OrderError | null {
    const result = this.paymentGateway.charge(userId, total);
    if (!result.success) {
      return {
        type: 'PAYMENT_FAILED',
        message: result.error || 'Payment processing failed',
      };
    }
    return null;
  }
}

// ============================================================================
// Notification Service
// ============================================================================

class OrderNotificationService {
  constructor(private readonly emailService: EmailService) {}

  notifyOrderConfirmed(email: string, orderId: string): void {
    this.emailService.sendOrderConfirmation(email, orderId);
  }
}

// ============================================================================
// Order Processor (Orchestrator)
// ============================================================================

class OrderProcessor {
  private readonly validator: OrderValidator;
  private readonly calculator: OrderCalculator;
  private readonly stockChecker: StockChecker;
  private readonly balanceChecker: BalanceChecker;
  private readonly paymentProcessor: PaymentProcessor;
  private readonly notificationService: OrderNotificationService;

  constructor(
    inventory: Inventory,
    paymentGateway: PaymentGateway,
    emailService: EmailService
  ) {
    this.validator = new OrderValidator();
    this.calculator = new OrderCalculator();
    this.stockChecker = new StockChecker(inventory);
    this.balanceChecker = new BalanceChecker();
    this.paymentProcessor = new PaymentProcessor(paymentGateway);
    this.notificationService = new OrderNotificationService(emailService);
  }

  processOrder(order: Order | null | undefined, user: User | null | undefined): OrderResult {
    // Step 1: Validate inputs
    const orderError = this.validator.validateOrder(order);
    if (orderError) {
      return { success: false, error: orderError };
    }

    const userError = this.validator.validateUser(user);
    if (userError) {
      return { success: false, error: userError };
    }

    // TypeScript now knows these are not null
    const validOrder = order!;
    const validUser = user!;

    // Step 2: Check stock availability
    const stockError = this.stockChecker.checkAvailability(validOrder.items);
    if (stockError) {
      return { success: false, error: stockError };
    }

    // Step 3: Calculate total
    const total = this.calculator.calculateTotal(validOrder.items);

    // Step 4: Check user balance
    const balanceError = this.balanceChecker.checkBalance(validUser, total);
    if (balanceError) {
      return { success: false, error: balanceError };
    }

    // Step 5: Process payment
    const paymentError = this.paymentProcessor.processPayment(validUser.id, total);
    if (paymentError) {
      return { success: false, error: paymentError };
    }

    // Step 6: Reserve inventory
    this.stockChecker.reserveItems(validOrder.items);

    // Step 7: Send notification
    this.notificationService.notifyOrderConfirmed(validUser.email, validOrder.id);

    // Step 8: Return success
    return {
      success: true,
      orderId: validOrder.id,
      total,
    };
  }
}

// ============================================================================
// Functional alternative using pipe
// ============================================================================

function processOrderFunctional(
  order: Order | null | undefined,
  user: User | null | undefined,
  inventory: Inventory,
  paymentGateway: PaymentGateway,
  emailService: EmailService
): OrderResult {
  // Validation
  if (!order || !user) {
    return { success: false, error: { type: 'INVALID_INPUT', message: 'Invalid input' } };
  }
  if (order.items.length === 0) {
    return { success: false, error: { type: 'EMPTY_CART', message: 'Cart is empty' } };
  }

  // Stock check
  const outOfStockItem = order.items.find(
    item => inventory.checkStock(item.id) < item.qty
  );
  if (outOfStockItem) {
    return {
      success: false,
      error: {
        type: 'OUT_OF_STOCK',
        itemName: outOfStockItem.name,
        message: `Item '${outOfStockItem.name}' is out of stock`,
      },
    };
  }

  // Calculate total
  const total = order.items.reduce((sum, item) => sum + item.price * item.qty, 0);

  // Balance check
  if (user.balance < total) {
    return {
      success: false,
      error: {
        type: 'INSUFFICIENT_FUNDS',
        required: total,
        available: user.balance,
        message: 'Insufficient funds',
      },
    };
  }

  // Payment
  const payment = paymentGateway.charge(user.id, total);
  if (!payment.success) {
    return {
      success: false,
      error: { type: 'PAYMENT_FAILED', message: payment.error || 'Payment failed' },
    };
  }

  // Reserve inventory
  order.items.forEach(item => inventory.decreaseStock(item.id, item.qty));

  // Notification
  emailService.sendOrderConfirmation(user.email, order.id);

  return { success: true, orderId: order.id, total };
}

// ============================================================================
// Tests
// ============================================================================

function runTests(): void {
  const results: { name: string; passed: boolean; error?: string }[] = [];

  function test(name: string, fn: () => void): void {
    try {
      fn();
      results.push({ name, passed: true });
    } catch (error) {
      results.push({ name, passed: false, error: String(error) });
    }
  }

  function assertEqual<T>(actual: T, expected: T, message?: string): void {
    if (actual !== expected) {
      throw new Error(message || `Expected ${expected}, got ${actual}`);
    }
  }

  // Mock implementations
  function createMockInventory(stock: Record<string, number>): Inventory & { decreasedItems: Array<{ id: string; qty: number }> } {
    const decreasedItems: Array<{ id: string; qty: number }> = [];
    return {
      checkStock: (id: string) => stock[id] || 0,
      decreaseStock: (id: string, qty: number) => {
        decreasedItems.push({ id, qty });
        stock[id] = (stock[id] || 0) - qty;
      },
      decreasedItems,
    };
  }

  function createMockPaymentGateway(shouldSucceed: boolean): PaymentGateway & { chargedAmounts: number[] } {
    const chargedAmounts: number[] = [];
    return {
      charge: (_userId: string, amount: number) => {
        chargedAmounts.push(amount);
        return shouldSucceed
          ? { success: true, transactionId: 'tx-123' }
          : { success: false, error: 'Card declined' };
      },
      chargedAmounts,
    };
  }

  function createMockEmailService(): EmailService & { sentEmails: Array<{ email: string; orderId: string }> } {
    const sentEmails: Array<{ email: string; orderId: string }> = [];
    return {
      sendOrderConfirmation: (email: string, orderId: string) => {
        sentEmails.push({ email, orderId });
      },
      sentEmails,
    };
  }

  // Test fixtures
  const validOrder: Order = {
    id: 'order-123',
    items: [
      { id: 'item-1', name: 'Widget', price: 10, qty: 2 },
      { id: 'item-2', name: 'Gadget', price: 20, qty: 1 },
    ],
  };

  const validUser: User = {
    id: 'user-123',
    email: 'test@example.com',
    balance: 100,
  };

  // Tests for OrderProcessor class
  test('should process order successfully', () => {
    const inventory = createMockInventory({ 'item-1': 10, 'item-2': 5 });
    const payment = createMockPaymentGateway(true);
    const email = createMockEmailService();
    const processor = new OrderProcessor(inventory, payment, email);

    const result = processor.processOrder(validOrder, validUser);

    assertEqual(result.success, true);
    if (result.success) {
      assertEqual(result.orderId, 'order-123');
      assertEqual(result.total, 40);
    }
  });

  test('should fail with null order', () => {
    const inventory = createMockInventory({});
    const payment = createMockPaymentGateway(true);
    const email = createMockEmailService();
    const processor = new OrderProcessor(inventory, payment, email);

    const result = processor.processOrder(null, validUser);

    assertEqual(result.success, false);
    if (!result.success) {
      assertEqual(result.error.type, 'INVALID_INPUT');
    }
  });

  test('should fail with null user', () => {
    const inventory = createMockInventory({});
    const payment = createMockPaymentGateway(true);
    const email = createMockEmailService();
    const processor = new OrderProcessor(inventory, payment, email);

    const result = processor.processOrder(validOrder, null);

    assertEqual(result.success, false);
    if (!result.success) {
      assertEqual(result.error.type, 'INVALID_INPUT');
    }
  });

  test('should fail with empty cart', () => {
    const inventory = createMockInventory({});
    const payment = createMockPaymentGateway(true);
    const email = createMockEmailService();
    const processor = new OrderProcessor(inventory, payment, email);

    const emptyOrder: Order = { id: 'order-123', items: [] };
    const result = processor.processOrder(emptyOrder, validUser);

    assertEqual(result.success, false);
    if (!result.success) {
      assertEqual(result.error.type, 'EMPTY_CART');
    }
  });

  test('should fail when item out of stock', () => {
    const inventory = createMockInventory({ 'item-1': 1, 'item-2': 5 }); // Only 1 of item-1
    const payment = createMockPaymentGateway(true);
    const email = createMockEmailService();
    const processor = new OrderProcessor(inventory, payment, email);

    const result = processor.processOrder(validOrder, validUser);

    assertEqual(result.success, false);
    if (!result.success) {
      assertEqual(result.error.type, 'OUT_OF_STOCK');
      if (result.error.type === 'OUT_OF_STOCK') {
        assertEqual(result.error.itemName, 'Widget');
      }
    }
  });

  test('should fail when insufficient funds', () => {
    const inventory = createMockInventory({ 'item-1': 10, 'item-2': 5 });
    const payment = createMockPaymentGateway(true);
    const email = createMockEmailService();
    const processor = new OrderProcessor(inventory, payment, email);

    const poorUser: User = { ...validUser, balance: 30 };
    const result = processor.processOrder(validOrder, poorUser);

    assertEqual(result.success, false);
    if (!result.success) {
      assertEqual(result.error.type, 'INSUFFICIENT_FUNDS');
    }
  });

  test('should fail when payment fails', () => {
    const inventory = createMockInventory({ 'item-1': 10, 'item-2': 5 });
    const payment = createMockPaymentGateway(false);
    const email = createMockEmailService();
    const processor = new OrderProcessor(inventory, payment, email);

    const result = processor.processOrder(validOrder, validUser);

    assertEqual(result.success, false);
    if (!result.success) {
      assertEqual(result.error.type, 'PAYMENT_FAILED');
    }
  });

  test('should decrease inventory after successful order', () => {
    const inventory = createMockInventory({ 'item-1': 10, 'item-2': 5 });
    const payment = createMockPaymentGateway(true);
    const email = createMockEmailService();
    const processor = new OrderProcessor(inventory, payment, email);

    processor.processOrder(validOrder, validUser);

    assertEqual(inventory.decreasedItems.length, 2);
    assertEqual(inventory.decreasedItems[0].id, 'item-1');
    assertEqual(inventory.decreasedItems[0].qty, 2);
  });

  test('should send email after successful order', () => {
    const inventory = createMockInventory({ 'item-1': 10, 'item-2': 5 });
    const payment = createMockPaymentGateway(true);
    const email = createMockEmailService();
    const processor = new OrderProcessor(inventory, payment, email);

    processor.processOrder(validOrder, validUser);

    assertEqual(email.sentEmails.length, 1);
    assertEqual(email.sentEmails[0].email, 'test@example.com');
    assertEqual(email.sentEmails[0].orderId, 'order-123');
  });

  test('should not decrease inventory when payment fails', () => {
    const inventory = createMockInventory({ 'item-1': 10, 'item-2': 5 });
    const payment = createMockPaymentGateway(false);
    const email = createMockEmailService();
    const processor = new OrderProcessor(inventory, payment, email);

    processor.processOrder(validOrder, validUser);

    assertEqual(inventory.decreasedItems.length, 0);
  });

  test('should not send email when payment fails', () => {
    const inventory = createMockInventory({ 'item-1': 10, 'item-2': 5 });
    const payment = createMockPaymentGateway(false);
    const email = createMockEmailService();
    const processor = new OrderProcessor(inventory, payment, email);

    processor.processOrder(validOrder, validUser);

    assertEqual(email.sentEmails.length, 0);
  });

  // Print results
  console.log('\n=== Test Results ===\n');
  let passed = 0;
  let failed = 0;
  for (const result of results) {
    if (result.passed) {
      console.log(`✓ ${result.name}`);
      passed++;
    } else {
      console.log(`✗ ${result.name}`);
      console.log(`  Error: ${result.error}`);
      failed++;
    }
  }
  console.log(`\nTotal: ${passed} passed, ${failed} failed`);
}

runTests();

File: docs/comparison-tests/results-with-mcp/python/test-4-result.py
"""
Test PY-4: Refactor
Refactor to Python best practices.

Original code had issues:
- Dictionary-based data instead of dataclasses
- No type hints
- C-style loops (range(len(...)))
- Boolean comparison with == False
- String concatenation instead of f-strings
- No proper error handling

Apply: dataclasses, type hints, protocols, Pythonic idioms.
"""

from __future__ import annotations

from abc import abstractmethod
from dataclasses import dataclass, field
from decimal import Decimal
from enum import Enum, auto
from typing import Protocol


# ============================================================================
# Custom Exceptions
# ============================================================================

class OrderError(Exception):
    """Base exception for order errors."""
    pass


class InvalidInputError(OrderError):
    """Raised when input is invalid."""
    pass


class EmptyCartError(OrderError):
    """Raised when cart is empty."""
    pass


class OutOfStockError(OrderError):
    """Raised when item is out of stock."""

    def __init__(self, item_name: str) -> None:
        self.item_name = item_name
        super().__init__(f"Item '{item_name}' is out of stock")


class InsufficientFundsError(OrderError):
    """Raised when user has insufficient funds."""

    def __init__(self, required: Decimal, available: Decimal) -> None:
        self.required = required
        self.available = available
        super().__init__(f"Insufficient funds: required {required}, available {available}")


class PaymentFailedError(OrderError):
    """Raised when payment processing fails."""
    pass


# ============================================================================
# Domain Types (Value Objects and Entities)
# ============================================================================

@dataclass(frozen=True)
class OrderItem:
    """Immutable order item."""
    id: str
    name: str
    price: Decimal
    quantity: int

    @property
    def subtotal(self) -> Decimal:
        """Calculate item subtotal."""
        return self.price * self.quantity


@dataclass(frozen=True)
class Order:
    """Immutable order entity."""
    id: str
    items: tuple[OrderItem, ...]

    @property
    def total(self) -> Decimal:
        """Calculate order total."""
        return sum((item.subtotal for item in self.items), Decimal(0))

    @property
    def is_empty(self) -> bool:
        """Check if order has no items."""
        return len(self.items) == 0


@dataclass(frozen=True)
class User:
    """User entity."""
    id: str
    email: str
    balance: Decimal


@dataclass(frozen=True)
class PaymentResult:
    """Result of a payment operation."""
    success: bool
    transaction_id: str | None = None
    error_message: str | None = None


class OrderStatus(Enum):
    """Order processing status."""
    SUCCESS = auto()
    INVALID_INPUT = auto()
    EMPTY_CART = auto()
    OUT_OF_STOCK = auto()
    INSUFFICIENT_FUNDS = auto()
    PAYMENT_FAILED = auto()


@dataclass(frozen=True)
class OrderResult:
    """Result of order processing."""
    success: bool
    status: OrderStatus
    order_id: str | None = None
    total: Decimal | None = None
    error_message: str | None = None

    @classmethod
    def successful(cls, order_id: str, total: Decimal) -> OrderResult:
        """Create a successful result."""
        return cls(
            success=True,
            status=OrderStatus.SUCCESS,
            order_id=order_id,
            total=total,
        )

    @classmethod
    def failed(cls, status: OrderStatus, message: str) -> OrderResult:
        """Create a failed result."""
        return cls(
            success=False,
            status=status,
            error_message=message,
        )


# ============================================================================
# Protocol Interfaces (Dependency Inversion)
# ============================================================================

class Inventory(Protocol):
    """Protocol for inventory operations."""

    @abstractmethod
    def check_stock(self, item_id: str) -> int:
        """Check available stock for an item."""
        ...

    @abstractmethod
    def decrease_stock(self, item_id: str, quantity: int) -> None:
        """Decrease stock for an item."""
        ...


class PaymentGateway(Protocol):
    """Protocol for payment processing."""

    @abstractmethod
    def charge(self, user_id: str, amount: Decimal) -> PaymentResult:
        """Charge a user."""
        ...


class EmailService(Protocol):
    """Protocol for email notifications."""

    @abstractmethod
    def send(self, to: str, subject: str, body: str) -> None:
        """Send an email."""
        ...


# ============================================================================
# Order Validator
# ============================================================================

class OrderValidator:
    """Validates orders before processing."""

    def validate(self, order: Order | None, user: User | None) -> OrderResult | None:
        """
        Validate order and user.
        Returns None if valid, otherwise returns an error result.
        """
        if order is None or user is None:
            return OrderResult.failed(
                OrderStatus.INVALID_INPUT,
                "Order and user are required",
            )

        if order.is_empty:
            return OrderResult.failed(
                OrderStatus.EMPTY_CART,
                "Cart is empty",
            )

        return None


# ============================================================================
# Stock Checker
# ============================================================================

class StockChecker:
    """Checks and manages inventory stock."""

    def __init__(self, inventory: Inventory) -> None:
        self._inventory = inventory

    def check_availability(self, items: tuple[OrderItem, ...]) -> str | None:
        """
        Check if all items are in stock.
        Returns item name if out of stock, None if all available.
        """
        for item in items:
            available = self._inventory.check_stock(item.id)
            if available < item.quantity:
                return item.name
        return None

    def reserve_items(self, items: tuple[OrderItem, ...]) -> None:
        """Reserve items by decreasing stock."""
        for item in items:
            self._inventory.decrease_stock(item.id, item.quantity)


# ============================================================================
# Refactored Order Processor
# ============================================================================

class OrderProcessor:
    """Processes orders with proper validation and error handling."""

    def __init__(
        self,
        inventory: Inventory,
        payment_gateway: PaymentGateway,
        email_service: EmailService,
    ) -> None:
        self._inventory = inventory
        self._payment_gateway = payment_gateway
        self._email_service = email_service
        self._validator = OrderValidator()
        self._stock_checker = StockChecker(inventory)

    def process_order(self, order: Order | None, user: User | None) -> OrderResult:
        """
        Process an order.

        This is the refactored version with:
        - Type hints
        - Dataclasses instead of dicts
        - Pythonic iteration
        - Proper exception handling
        - Single responsibility (delegating to specialized classes)
        """
        # Validate inputs
        validation_error = self._validator.validate(order, user)
        if validation_error:
            return validation_error

        # Type narrowing - at this point we know order and user are not None
        assert order is not None
        assert user is not None

        # Check stock availability
        out_of_stock_item = self._stock_checker.check_availability(order.items)
        if out_of_stock_item:
            return OrderResult.failed(
                OrderStatus.OUT_OF_STOCK,
                f"Out of stock: {out_of_stock_item}",
            )

        # Calculate total
        total = order.total

        # Check user balance
        if user.balance < total:
            return OrderResult.failed(
                OrderStatus.INSUFFICIENT_FUNDS,
                f"Insufficient funds: required {total}, available {user.balance}",
            )

        # Process payment
        payment_result = self._payment_gateway.charge(user.id, total)
        if not payment_result.success:
            return OrderResult.failed(
                OrderStatus.PAYMENT_FAILED,
                payment_result.error_message or "Payment processing failed",
            )

        # Reserve inventory
        self._stock_checker.reserve_items(order.items)

        # Send confirmation email
        self._send_confirmation(user.email, order.id)

        return OrderResult.successful(order.id, total)

    def _send_confirmation(self, email: str, order_id: str) -> None:
        """Send order confirmation email."""
        self._email_service.send(
            to=email,
            subject="Order confirmed",
            body=f"Your order #{order_id} is confirmed.",
        )


# ============================================================================
# Mock Implementations for Testing
# ============================================================================

@dataclass
class MockInventory:
    """Mock inventory for testing."""
    stock: dict[str, int] = field(default_factory=dict)
    decreased: list[tuple[str, int]] = field(default_factory=list)

    def check_stock(self, item_id: str) -> int:
        return self.stock.get(item_id, 0)

    def decrease_stock(self, item_id: str, quantity: int) -> None:
        self.decreased.append((item_id, quantity))
        self.stock[item_id] = self.stock.get(item_id, 0) - quantity


@dataclass
class MockPaymentGateway:
    """Mock payment gateway for testing."""
    should_succeed: bool = True
    charged: list[tuple[str, Decimal]] = field(default_factory=list)

    def charge(self, user_id: str, amount: Decimal) -> PaymentResult:
        self.charged.append((user_id, amount))
        if self.should_succeed:
            return PaymentResult(success=True, transaction_id="tx-123")
        return PaymentResult(success=False, error_message="Card declined")


@dataclass
class MockEmailService:
    """Mock email service for testing."""
    sent: list[tuple[str, str, str]] = field(default_factory=list)

    def send(self, to: str, subject: str, body: str) -> None:
        self.sent.append((to, subject, body))


# ============================================================================
# Tests
# ============================================================================

import pytest


@pytest.fixture
def inventory() -> MockInventory:
    """Create mock inventory with default stock."""
    return MockInventory(stock={"item-1": 10, "item-2": 5})


@pytest.fixture
def payment_gateway() -> MockPaymentGateway:
    """Create mock payment gateway."""
    return MockPaymentGateway(should_succeed=True)


@pytest.fixture
def email_service() -> MockEmailService:
    """Create mock email service."""
    return MockEmailService()


@pytest.fixture
def processor(
    inventory: MockInventory,
    payment_gateway: MockPaymentGateway,
    email_service: MockEmailService,
) -> OrderProcessor:
    """Create order processor with mock dependencies."""
    return OrderProcessor(inventory, payment_gateway, email_service)


@pytest.fixture
def valid_order() -> Order:
    """Create a valid test order."""
    return Order(
        id="order-123",
        items=(
            OrderItem(id="item-1", name="Widget", price=Decimal("10"), quantity=2),
            OrderItem(id="item-2", name="Gadget", price=Decimal("20"), quantity=1),
        ),
    )


@pytest.fixture
def valid_user() -> User:
    """Create a valid test user."""
    return User(id="user-123", email="test@example.com", balance=Decimal("100"))


class TestOrderProcessor:
    """Tests for OrderProcessor."""

    def test_successful_order(
        self,
        processor: OrderProcessor,
        valid_order: Order,
        valid_user: User,
        email_service: MockEmailService,
    ) -> None:
        """Should process order successfully."""
        result = processor.process_order(valid_order, valid_user)

        assert result.success is True
        assert result.status == OrderStatus.SUCCESS
        assert result.order_id == "order-123"
        assert result.total == Decimal("40")  # 10*2 + 20*1
        assert len(email_service.sent) == 1

    def test_fails_with_none_order(self, processor: OrderProcessor, valid_user: User) -> None:
        """Should fail when order is None."""
        result = processor.process_order(None, valid_user)

        assert result.success is False
        assert result.status == OrderStatus.INVALID_INPUT

    def test_fails_with_none_user(self, processor: OrderProcessor, valid_order: Order) -> None:
        """Should fail when user is None."""
        result = processor.process_order(valid_order, None)

        assert result.success is False
        assert result.status == OrderStatus.INVALID_INPUT

    def test_fails_with_empty_cart(self, processor: OrderProcessor, valid_user: User) -> None:
        """Should fail when cart is empty."""
        empty_order = Order(id="order-123", items=())

        result = processor.process_order(empty_order, valid_user)

        assert result.success is False
        assert result.status == OrderStatus.EMPTY_CART

    def test_fails_when_out_of_stock(
        self,
        inventory: MockInventory,
        payment_gateway: MockPaymentGateway,
        email_service: MockEmailService,
        valid_user: User,
    ) -> None:
        """Should fail when item is out of stock."""
        inventory.stock["item-1"] = 1  # Only 1 in stock, but order needs 2
        processor = OrderProcessor(inventory, payment_gateway, email_service)
        order = Order(
            id="order-123",
            items=(OrderItem(id="item-1", name="Widget", price=Decimal("10"), quantity=2),),
        )

        result = processor.process_order(order, valid_user)

        assert result.success is False
        assert result.status == OrderStatus.OUT_OF_STOCK
        assert "Widget" in (result.error_message or "")

    def test_fails_with_insufficient_funds(
        self,
        processor: OrderProcessor,
        valid_order: Order,
    ) -> None:
        """Should fail when user has insufficient funds."""
        poor_user = User(id="user-123", email="test@example.com", balance=Decimal("30"))

        result = processor.process_order(valid_order, poor_user)

        assert result.success is False
        assert result.status == OrderStatus.INSUFFICIENT_FUNDS

    def test_fails_when_payment_fails(
        self,
        inventory: MockInventory,
        email_service: MockEmailService,
        valid_order: Order,
        valid_user: User,
    ) -> None:
        """Should fail when payment fails."""
        failed_payment = MockPaymentGateway(should_succeed=False)
        processor = OrderProcessor(inventory, failed_payment, email_service)

        result = processor.process_order(valid_order, valid_user)

        assert result.success is False
        assert result.status == OrderStatus.PAYMENT_FAILED

    def test_decreases_inventory_on_success(
        self,
        processor: OrderProcessor,
        valid_order: Order,
        valid_user: User,
        inventory: MockInventory,
    ) -> None:
        """Should decrease inventory on successful order."""
        processor.process_order(valid_order, valid_user)

        assert ("item-1", 2) in inventory.decreased
        assert ("item-2", 1) in inventory.decreased

    def test_does_not_decrease_inventory_on_payment_failure(
        self,
        inventory: MockInventory,
        email_service: MockEmailService,
        valid_order: Order,
        valid_user: User,
    ) -> None:
        """Should not decrease inventory when payment fails."""
        failed_payment = MockPaymentGateway(should_succeed=False)
        processor = OrderProcessor(inventory, failed_payment, email_service)

        processor.process_order(valid_order, valid_user)

        assert len(inventory.decreased) == 0

    def test_sends_email_on_success(
        self,
        processor: OrderProcessor,
        valid_order: Order,
        valid_user: User,
        email_service: MockEmailService,
    ) -> None:
        """Should send confirmation email on success."""
        processor.process_order(valid_order, valid_user)

        assert len(email_service.sent) == 1
        to, subject, body = email_service.sent[0]
        assert to == "test@example.com"
        assert "confirmed" in subject.lower()
        assert "order-123" in body


class TestOrderItem:
    """Tests for OrderItem dataclass."""

    def test_subtotal_calculation(self) -> None:
        """Should calculate subtotal correctly."""
        item = OrderItem(id="1", name="Test", price=Decimal("10.50"), quantity=3)

        assert item.subtotal == Decimal("31.50")

    def test_is_immutable(self) -> None:
        """OrderItem should be immutable."""
        item = OrderItem(id="1", name="Test", price=Decimal("10"), quantity=1)

        with pytest.raises(Exception):  # FrozenInstanceError
            item.quantity = 5  # type: ignore


class TestOrder:
    """Tests for Order dataclass."""

    def test_total_calculation(self) -> None:
        """Should calculate total correctly."""
        order = Order(
            id="1",
            items=(
                OrderItem(id="1", name="A", price=Decimal("10"), quantity=2),
                OrderItem(id="2", name="B", price=Decimal("5"), quantity=4),
            ),
        )

        assert order.total == Decimal("40")  # 20 + 20

    def test_is_empty(self) -> None:
        """Should detect empty order."""
        empty = Order(id="1", items=())
        non_empty = Order(id="2", items=(OrderItem(id="1", name="A", price=Decimal("10"), quantity=1),))

        assert empty.is_empty is True
        assert non_empty.is_empty is False


class TestOrderResult:
    """Tests for OrderResult dataclass."""

    def test_successful_factory(self) -> None:
        """Should create successful result."""
        result = OrderResult.successful("order-1", Decimal("100"))

        assert result.success is True
        assert result.status == OrderStatus.SUCCESS
        assert result.order_id == "order-1"
        assert result.total == Decimal("100")

    def test_failed_factory(self) -> None:
        """Should create failed result."""
        result = OrderResult.failed(OrderStatus.OUT_OF_STOCK, "Widget out of stock")

        assert result.success is False
        assert result.status == OrderStatus.OUT_OF_STOCK
        assert result.error_message == "Widget out of stock"


if __name__ == "__main__":
    pytest.main([__file__, "-v"])
